File: Telerik/SaveLoadSettings/RadGridViewSettings.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Configuration;
using System.ComponentModel;
using System.Windows;
using System.Windows.Data;
using Telerik.Windows.Controls;
using Telerik.Windows.Data;
using Telerik.Windows.Controls.GridView;
using System.Windows.Controls;
using System.Runtime.Serialization;
using System.IO;
using System.IO.IsolatedStorage;

namespace Telerik.Windows.Controls.GridView.Settings
{
    public class RadGridViewSettings
    {
        public RadGridViewSettings()
        {
            //
        }

        public class RadGridViewApplicationSettings : Dictionary<string, object>
        {
            private RadGridViewSettings settings;

            private DataContractSerializer serializer = null;

            public RadGridViewApplicationSettings()
            {
                //
            }

            public RadGridViewApplicationSettings(RadGridViewSettings settings)
            {
                this.settings = settings;

                List<Type> types = new List<Type>();
                types.Add(typeof(List<ColumnSetting>));
                types.Add(typeof(List<FilterSetting>));
                types.Add(typeof(List<GroupSetting>));
                types.Add(typeof(List<SortSetting>));
                types.Add(typeof(List<PropertySetting>));

                this.serializer = new DataContractSerializer(typeof(RadGridViewApplicationSettings), types);
            }

            public string PersistID
            {
                get
                {
                    if (!ContainsKey("PersistID") && settings.grid != null)
                    {
                        this["PersistID"] = settings.grid.Name;
                    }

                    return (string) this["PersistID"];
                }
            }

            public int FrozenColumnCount
            {
                get
                {
                    if (!ContainsKey("FrozenColumnCount"))
                    {
                        this["FrozenColumnCount"] = 0;
                    }

                    return (int) this["FrozenColumnCount"];
                }
                set
                {
                    this["FrozenColumnCount"] = value;
                }
            }

            public List<ColumnSetting> ColumnSettings
            {
                get
                {
                    if (!ContainsKey("ColumnSettings"))
                    {
                        this["ColumnSettings"] = new List<ColumnSetting>();
                    }

                    return (List<ColumnSetting>) this["ColumnSettings"];
                }
            }

            public List<SortSetting> SortSettings
            {
                get
                {
                    if (!ContainsKey("SortSettings"))
                    {
                        this["SortSettings"] = new List<SortSetting>();
                    }

                    return (List<SortSetting>) this["SortSettings"];
                }
            }

            public List<GroupSetting> GroupSettings
            {
                get
                {
                    if (!ContainsKey("GroupSettings"))
                    {
                        this["GroupSettings"] = new List<GroupSetting>();
                    }

                    return (List<GroupSetting>) this["GroupSettings"];
                }
            }

            public List<FilterSetting> FilterSettings
            {
                get
                {
                    if (!ContainsKey("FilterSettings"))
                    {
                        this["FilterSettings"] = new List<FilterSetting>();
                    }

                    return (List<FilterSetting>) this["FilterSettings"];
                }
            }

            public void Reload()
            {
                try
                {
                    using (IsolatedStorageFile file = IsolatedStorageFile.GetUserStoreForApplication())
                    {
                        using (IsolatedStorageFileStream stream = new IsolatedStorageFileStream(PersistID, FileMode.Open, file))
                        {
                            if (stream.Length > 0)
                            {
                                RadGridViewApplicationSettings loaded = (RadGridViewApplicationSettings) serializer.ReadObject(stream);

                                FrozenColumnCount = loaded.FrozenColumnCount;

                                ColumnSettings.Clear();
                                foreach (ColumnSetting cs in loaded.ColumnSettings)
                                {
                                    ColumnSettings.Add(cs);
                                }

                                FilterSettings.Clear();
                                foreach (FilterSetting fs in loaded.FilterSettings)
                                {
                                    FilterSettings.Add(fs);
                                }

                                GroupSettings.Clear();
                                foreach (GroupSetting gs in loaded.GroupSettings)
                                {
                                    GroupSettings.Add(gs);
                                }

                                SortSettings.Clear();
                                foreach (SortSetting ss in loaded.SortSettings)
                                {
                                    SortSettings.Add(ss);
                                }
                            }
                        }
                    }
                }
                catch
                {

                }
            }

            public void Reset()
            {
                try
                {
                    using (IsolatedStorageFile file = IsolatedStorageFile.GetUserStoreForApplication())
                    {
                        file.DeleteFile(PersistID);
                    }
                }
                catch
                {
                    //
                }
            }

            public void Save()
            {
                try
                {
                    using (IsolatedStorageFile file = IsolatedStorageFile.GetUserStoreForApplication())
                    {
                        using (IsolatedStorageFileStream stream = new IsolatedStorageFileStream(PersistID, FileMode.Create, file))
                        {
                            serializer.WriteObject(stream, this);
                        }
                    }
                }
                catch
                {
                    //
                }
            }
        }

        private RadGridView grid = null;

        public RadGridViewSettings(RadGridView grid)
        {
            this.grid = grid;
        }

        public static readonly DependencyProperty IsEnabledProperty
           = DependencyProperty.RegisterAttached("IsEnabled", typeof(bool), typeof(RadGridViewSettings),
                new PropertyMetadata(new PropertyChangedCallback(OnIsEnabledPropertyChanged)));

        public static bool GetIsEnabled(DependencyObject dependencyObject)
        {
            return (bool) dependencyObject.GetValue(IsEnabledProperty);
        }

        public static void SetIsEnabled(DependencyObject dependencyObject, bool enabled)
        {
            dependencyObject.SetValue(IsEnabledProperty, enabled);
        }

        private static void OnIsEnabledPropertyChanged(DependencyObject dependencyObject, DependencyPropertyChangedEventArgs e)
        {
            RadGridView grid = dependencyObject as RadGridView;
            if (grid != null)
            {
                if ((bool) e.NewValue)
                {
                    RadGridViewSettings settings = new RadGridViewSettings(grid);
                    settings.Attach();
                }
            }
        }

        public virtual void LoadState()
        {
            try
            {
                Settings.Reload();
            }
            catch
            {
                Settings.Reset();
            }

            if (this.grid != null)
            {
                grid.FrozenColumnCount = Settings.FrozenColumnCount;

                if (Settings.ColumnSettings.Count > 0)
                {
                    foreach (ColumnSetting setting in Settings.ColumnSettings)
                    {
                        ColumnSetting currentSetting = setting;

                        GridViewDataColumn column = (from c in grid.Columns.OfType<GridViewDataColumn>()
                                                     where c.UniqueName == currentSetting.UniqueName
                                                     select c).FirstOrDefault();

                        if (column != null)
                        {
                            if (currentSetting.DisplayIndex != null)
                            {
                                column.DisplayIndex = currentSetting.DisplayIndex.Value;
                            }

                            if (setting.Width != null)
                            {
                                column.Width = new GridViewLength(setting.Width.Value);
                            }
                        }
                    }
                }
                using (grid.DeferRefresh())
                {
                    if (Settings.SortSettings.Count > 0)
                    {
                        grid.SortDescriptors.Clear();

                        foreach (SortSetting setting in Settings.SortSettings)
                        {
                            Telerik.Windows.Data.SortDescriptor d = new Telerik.Windows.Data.SortDescriptor();
                            d.Member = setting.PropertyName;
                            d.SortDirection = setting.SortDirection;

                            grid.SortDescriptors.Add(d);
                        }
                    }

                    if (Settings.GroupSettings.Count > 0)
                    {
                        grid.GroupDescriptors.Clear();

                        foreach (GroupSetting setting in Settings.GroupSettings)
                        {
                            Telerik.Windows.Data.GroupDescriptor d = new Telerik.Windows.Data.GroupDescriptor();
                            d.Member = setting.PropertyName;
                            d.SortDirection = setting.SortDirection;
                            d.DisplayContent = setting.DisplayContent;

                            grid.GroupDescriptors.Add(d);
                        }
                    }

                    if (Settings.FilterSettings.Count > 0)
                    {
                        foreach (FilterSetting setting in Settings.FilterSettings)
                        {
                            FilterSetting currentSetting = setting;

                            GridViewDataColumn matchingColumn =
                            (from column in grid.Columns.OfType<GridViewDataColumn>()
                             where column.DataMemberBinding.Path.Path == currentSetting.PropertyName
                             select column).FirstOrDefault();

                            if (matchingColumn != null)
                            {
                                ColumnFilterDescriptor cfd = new ColumnFilterDescriptor(matchingColumn);

                                cfd.FieldFilter.Filter1.Member = setting.Filter1.Member;
                                cfd.FieldFilter.Filter1.Operator = setting.Filter1.Operator;
                                cfd.FieldFilter.Filter1.Value = setting.Filter1.Value;

                                cfd.FieldFilter.Filter2.Member = setting.Filter2.Member;
                                cfd.FieldFilter.Filter2.Operator = setting.Filter2.Operator;
                                cfd.FieldFilter.Filter2.Value = setting.Filter2.Value;

                                foreach (Telerik.Windows.Data.FilterDescriptor descriptor in setting.SelectedDistinctValues)
                                {
                                    cfd.DistinctFilter.FilterDescriptors.Add(descriptor);
                                }

                                this.grid.FilterDescriptors.Add(cfd);
                            }
                        }
                    }
                }
            }
        }

        public virtual void ResetState()
        {
            Settings.Reset();
        }

        public virtual void SaveState()
        {
            Settings.Reset();

            if (grid != null)
            {
                if (grid.Columns != null)
                {
                    Settings.ColumnSettings.Clear();

                    foreach (GridViewColumn column in grid.Columns)
                    {
                        if (column is GridViewDataColumn)
                        {
                            GridViewDataColumn dataColumn = (GridViewDataColumn) column;

                            ColumnSetting setting = new ColumnSetting();
                            setting.PropertyName = dataColumn.DataMemberBinding.Path.Path;
                            setting.UniqueName = dataColumn.UniqueName;
                            setting.Header = dataColumn.Header;
                            setting.Width = dataColumn.ActualWidth;
                            setting.DisplayIndex = dataColumn.DisplayIndex;
 
                            Settings.ColumnSettings.Add(setting);
                        }
                    }
                }

                if (grid.FilterDescriptors != null)
                {
                    Settings.FilterSettings.Clear();

                    foreach (IColumnFilterDescriptor cfd in grid.FilterDescriptors.OfType<IColumnFilterDescriptor>())
                    {
                        FilterSetting setting = new FilterSetting();

                        setting.Filter1 = new Telerik.Windows.Data.FilterDescriptor();
                        setting.Filter1.Member = cfd.FieldFilter.Filter1.Member;
                        setting.Filter1.Operator = cfd.FieldFilter.Filter1.Operator;
                        setting.Filter1.Value = cfd.FieldFilter.Filter1.Value;
                        setting.Filter1.MemberType = null;

                        setting.Filter2 = new Telerik.Windows.Data.FilterDescriptor();
                        setting.Filter2.Member = cfd.FieldFilter.Filter2.Member;
                        setting.Filter2.Operator = cfd.FieldFilter.Filter2.Operator;
                        setting.Filter2.Value = cfd.FieldFilter.Filter2.Value;
                        setting.Filter2.MemberType = null;

                        foreach (Telerik.Windows.Data.FilterDescriptor fd in cfd.DistinctFilter.FilterDescriptors.OfType<Telerik.Windows.Data.FilterDescriptor>())
                        {
                            Telerik.Windows.Data.FilterDescriptor newFd = new Telerik.Windows.Data.FilterDescriptor();
                            newFd.Member = fd.Member;
                            newFd.Operator = fd.Operator;
                            newFd.Value = fd.Value;
                            newFd.MemberType = null;
                            setting.SelectedDistinctValues.Add(newFd);
                        }

                        setting.PropertyName = cfd.Column.DataMemberBinding.Path.Path;

                        Settings.FilterSettings.Add(setting);
                    }
                }

                if (grid.SortDescriptors != null)
                {
                    Settings.SortSettings.Clear();

                    foreach (Telerik.Windows.Data.SortDescriptor d in grid.SortDescriptors)
                    {
                        SortSetting setting = new SortSetting();

                        setting.PropertyName = d.Member;
                        setting.SortDirection = d.SortDirection;

                        Settings.SortSettings.Add(setting);
                    }
                }

                if (grid.GroupDescriptors != null)
                {
                    Settings.GroupSettings.Clear();

                    foreach (Telerik.Windows.Data.GroupDescriptor d in grid.GroupDescriptors)
                    {
                        GroupSetting setting = new GroupSetting();

                        setting.PropertyName = d.Member;
                        setting.SortDirection = d.SortDirection;
                        setting.DisplayContent = d.DisplayContent;

                        Settings.GroupSettings.Add(setting);
                    }
                }

                Settings.FrozenColumnCount = grid.FrozenColumnCount;
            }

            Settings.Save();
        }

        private void Attach()
        {
            if (this.grid != null)
            {
                this.grid.LayoutUpdated += new EventHandler(LayoutUpdated);
                this.grid.Loaded += Loaded;
                Application.Current.Exit += Current_Exit;
            }
        }

        void Current_Exit(object sender, EventArgs e)
        {
            SaveState();
        }

        void Loaded(object sender, EventArgs e)
        {
            LoadState();
        }

        void LayoutUpdated(object sender, EventArgs e)
        {
            if (grid.Parent == null)
            {
                SaveState();
            }
        }

        private RadGridViewApplicationSettings gridViewApplicationSettings = null;

        protected virtual RadGridViewApplicationSettings CreateRadGridViewApplicationSettingsInstance()
        {
            return new RadGridViewApplicationSettings(this);
        }

        protected RadGridViewApplicationSettings Settings
        {
            get
            {
                if (gridViewApplicationSettings == null)
                {
                    gridViewApplicationSettings = CreateRadGridViewApplicationSettingsInstance();
                }
                return gridViewApplicationSettings;
            }
        }
    }

    public class PropertySetting
    {
        string _PropertyName;
        public string PropertyName 
        {
            get
            {
                return _PropertyName;
            }
            set
            {
                _PropertyName = value;
            }
        }
    }

    public class SortSetting : PropertySetting
    {
        ListSortDirection _SortDirection;
        public ListSortDirection SortDirection 
        {
            get
            {
                return _SortDirection;
            }
            set
            {
                _SortDirection = value;
            }
        }
    }

    public class GroupSetting : PropertySetting
    {
        object _DisplayContent;
        public object DisplayContent 
        {
            get
            {
                return _DisplayContent;
            }
            set
            {
                _DisplayContent = value;
            }
        }

        ListSortDirection? _SortDirection;
        public ListSortDirection? SortDirection 
        {
            get
            {
                return _SortDirection;
            }
            set
            {
                _SortDirection = value;
            }
        }
    }

    public class FilterSetting : PropertySetting
    {
        List<Telerik.Windows.Data.FilterDescriptor> _SelectedDistinctValues;
        public List<Telerik.Windows.Data.FilterDescriptor> SelectedDistinctValues 
        {
            get
            {
                if (_SelectedDistinctValues == null)
                {
                    _SelectedDistinctValues = new List<Telerik.Windows.Data.FilterDescriptor>();
                }
                return _SelectedDistinctValues;
            }
        }

        Telerik.Windows.Data.FilterDescriptor _Filter1;
        public Telerik.Windows.Data.FilterDescriptor Filter1 
        {
            get
            {
                return _Filter1;
            }
            set
            {
                _Filter1 = value;
            }
        }

        Telerik.Windows.Data.FilterDescriptor _Filter2;
        public Telerik.Windows.Data.FilterDescriptor Filter2 
        {
            get
            {
                return _Filter2;
            }
            set
            {
                _Filter2 = value;
            }
        }
    }

    public class ColumnSetting : PropertySetting
    {
        string _UniqueName;
        public string UniqueName 
        {
            get
            {
                return _UniqueName;
            }
            set
            {
                _UniqueName = value;
            }
        }

        object _Header;
        public object Header 
        {
            get
            {
                return _Header;
            }
            set
            {
                _Header = value;
            }
        }

        double? _Width;
        public double? Width 
        {
            get
            {
                return _Width;
            }
            set
            {
                _Width = value;
            }
        }

        int? _DisplayIndex;
        public int? DisplayIndex
        {
            get
            {
                return _DisplayIndex;
            }
            set
            {
                _DisplayIndex = value;
            }
        }
    }
}

File: Alice/Kamikaze 1m/Telerik/SaveLoadSettings/RadGridViewSettings.vb
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Configuration
Imports System.ComponentModel
Imports System.Windows
Imports System.Windows.Data
Imports Telerik.Windows.Controls
Imports Telerik.Windows.Data
Imports Telerik.Windows.Controls.GridView
Imports System.Windows.Controls
Imports System.Runtime.Serialization
Imports System.IO
Imports System.IO.IsolatedStorage

Namespace Telerik.Windows.Controls.GridView.Settings
	Public Class RadGridViewSettings
				'
		Public Sub New()
		End Sub

		Public Class RadGridViewApplicationSettings
			Inherits Dictionary(Of String, Object)
			Private settings As RadGridViewSettings

			Private serializer As DataContractSerializer = Nothing

					'
			Public Sub New()
			End Sub

			Public Sub New(settings As RadGridViewSettings)
				Me.settings = settings

				Dim types As New List(Of Type)()
				types.Add(GetType(List(Of ColumnSetting)))
				types.Add(GetType(List(Of FilterSetting)))
				types.Add(GetType(List(Of GroupSetting)))
				types.Add(GetType(List(Of SortSetting)))
				types.Add(GetType(List(Of PropertySetting)))

				Me.serializer = New DataContractSerializer(GetType(RadGridViewApplicationSettings), types)
			End Sub

			Public ReadOnly Property PersistID() As String
				Get
					If Not ContainsKey("PersistID") AndAlso settings.grid IsNot Nothing Then
						Me("PersistID") = settings.grid.Name
					End If

					Return DirectCast(Me("PersistID"), String)
				End Get
			End Property

			Public Property FrozenColumnCount() As Integer
				Get
					If Not ContainsKey("FrozenColumnCount") Then
						Me("FrozenColumnCount") = 0
					End If

					Return CInt(Me("FrozenColumnCount"))
				End Get
				Set
					Me("FrozenColumnCount") = value
				End Set
			End Property

			Public ReadOnly Property ColumnSettings() As List(Of ColumnSetting)
				Get
					If Not ContainsKey("ColumnSettings") Then
						Me("ColumnSettings") = New List(Of ColumnSetting)()
					End If

					Return DirectCast(Me("ColumnSettings"), List(Of ColumnSetting))
				End Get
			End Property

			Public ReadOnly Property SortSettings() As List(Of SortSetting)
				Get
					If Not ContainsKey("SortSettings") Then
						Me("SortSettings") = New List(Of SortSetting)()
					End If

					Return DirectCast(Me("SortSettings"), List(Of SortSetting))
				End Get
			End Property

			Public ReadOnly Property GroupSettings() As List(Of GroupSetting)
				Get
					If Not ContainsKey("GroupSettings") Then
						Me("GroupSettings") = New List(Of GroupSetting)()
					End If

					Return DirectCast(Me("GroupSettings"), List(Of GroupSetting))
				End Get
			End Property

			Public ReadOnly Property FilterSettings() As List(Of FilterSetting)
				Get
					If Not ContainsKey("FilterSettings") Then
						Me("FilterSettings") = New List(Of FilterSetting)()
					End If

					Return DirectCast(Me("FilterSettings"), List(Of FilterSetting))
				End Get
			End Property

			Public Sub Reload()
				Try
					Using file As IsolatedStorageFile = IsolatedStorageFile.GetUserStoreForApplication()
						Using stream As New IsolatedStorageFileStream(PersistID, FileMode.Open, file)
							If stream.Length > 0 Then
								Dim loaded As RadGridViewApplicationSettings = DirectCast(serializer.ReadObject(stream), RadGridViewApplicationSettings)

								FrozenColumnCount = loaded.FrozenColumnCount

								ColumnSettings.Clear()
								For Each cs As ColumnSetting In loaded.ColumnSettings
									ColumnSettings.Add(cs)
								Next

								FilterSettings.Clear()
								For Each fs As FilterSetting In loaded.FilterSettings
									FilterSettings.Add(fs)
								Next

								GroupSettings.Clear()
								For Each gs As GroupSetting In loaded.GroupSettings
									GroupSettings.Add(gs)
								Next

								SortSettings.Clear()
								For Each ss As SortSetting In loaded.SortSettings
									SortSettings.Add(ss)
								Next
							End If
						End Using
					End Using

				Catch
				End Try
			End Sub

			Public Sub Reset()
				Try
					Using file As IsolatedStorageFile = IsolatedStorageFile.GetUserStoreForApplication()
						file.DeleteFile(PersistID)
					End Using
						'
				Catch
				End Try
			End Sub

			Public Sub Save()
				Try
					Using file As IsolatedStorageFile = IsolatedStorageFile.GetUserStoreForApplication()
						Using stream As New IsolatedStorageFileStream(PersistID, FileMode.Create, file)
							serializer.WriteObject(stream, Me)
						End Using
					End Using
						'
				Catch
				End Try
			End Sub
		End Class

		Private grid As RadGridView = Nothing

		Public Sub New(grid As RadGridView)
			Me.grid = grid
		End Sub

		Public Shared ReadOnly IsEnabledProperty As DependencyProperty = DependencyProperty.RegisterAttached("IsEnabled", GetType(Boolean), GetType(RadGridViewSettings), New PropertyMetadata(New PropertyChangedCallback(AddressOf OnIsEnabledPropertyChanged)))

		Public Shared Function GetIsEnabled(dependencyObject As DependencyObject) As Boolean
			Return CBool(dependencyObject.GetValue(IsEnabledProperty))
		End Function

		Public Shared Sub SetIsEnabled(dependencyObject As DependencyObject, enabled As Boolean)
			dependencyObject.SetValue(IsEnabledProperty, enabled)
		End Sub

		Private Shared Sub OnIsEnabledPropertyChanged(dependencyObject As DependencyObject, e As DependencyPropertyChangedEventArgs)
			Dim grid As RadGridView = TryCast(dependencyObject, RadGridView)
			If grid IsNot Nothing Then
				If CBool(e.NewValue) Then
					Dim settings As New RadGridViewSettings(grid)
					settings.Attach()
				End If
			End If
		End Sub

		Public Overridable Sub LoadState()
			Try
				Settings.Reload()
			Catch
				Settings.Reset()
			End Try

			If Me.grid IsNot Nothing Then
				grid.FrozenColumnCount = Settings.FrozenColumnCount

				If Settings.ColumnSettings.Count > 0 Then
					For Each setting As ColumnSetting In Settings.ColumnSettings
						Dim currentSetting As ColumnSetting = setting

						Dim column As GridViewDataColumn = (From c In grid.Columns.OfType(Of GridViewDataColumn)() _
							Where c.UniqueName = currentSetting.UniqueName _
							Select c).FirstOrDefault()

						If column IsNot Nothing Then
							If currentSetting.DisplayIndex IsNot Nothing Then
								column.DisplayIndex = currentSetting.DisplayIndex.Value
							End If

							If setting.Width IsNot Nothing Then
								column.Width = New GridViewLength(setting.Width.Value)
							End If
						End If
					Next
				End If
				Using grid.DeferRefresh()
					If Settings.SortSettings.Count > 0 Then
						grid.SortDescriptors.Clear()

						For Each setting As SortSetting In Settings.SortSettings
							Dim d As New Telerik.Windows.Data.SortDescriptor()
							d.Member = setting.PropertyName
							d.SortDirection = setting.SortDirection

							grid.SortDescriptors.Add(d)
						Next
					End If

					If Settings.GroupSettings.Count > 0 Then
						grid.GroupDescriptors.Clear()

						For Each setting As GroupSetting In Settings.GroupSettings
							Dim d As New Telerik.Windows.Data.GroupDescriptor()
							d.Member = setting.PropertyName
							d.SortDirection = setting.SortDirection
							d.DisplayContent = setting.DisplayContent

							grid.GroupDescriptors.Add(d)
						Next
					End If

					If Settings.FilterSettings.Count > 0 Then
						For Each setting As FilterSetting In Settings.FilterSettings
							Dim currentSetting As FilterSetting = setting

							Dim matchingColumn As GridViewDataColumn = (From column In grid.Columns.OfType(Of GridViewDataColumn)() _
								Where column.DataMemberBinding.Path.Path = currentSetting.PropertyName _
								Select column).FirstOrDefault()

							If matchingColumn IsNot Nothing Then
								Dim cfd As New ColumnFilterDescriptor(matchingColumn)

								cfd.FieldFilter.Filter1.Member = setting.Filter1.Member
								cfd.FieldFilter.Filter1.[Operator] = setting.Filter1.[Operator]
								cfd.FieldFilter.Filter1.Value = setting.Filter1.Value

								cfd.FieldFilter.Filter2.Member = setting.Filter2.Member
								cfd.FieldFilter.Filter2.[Operator] = setting.Filter2.[Operator]
								cfd.FieldFilter.Filter2.Value = setting.Filter2.Value

								For Each descriptor As Telerik.Windows.Data.FilterDescriptor In setting.SelectedDistinctValues
									cfd.DistinctFilter.FilterDescriptors.Add(descriptor)
								Next

								Me.grid.FilterDescriptors.Add(cfd)
							End If
						Next
					End If
				End Using
			End If
		End Sub

		Public Overridable Sub ResetState()
			Settings.Reset()
		End Sub

		Public Overridable Sub SaveState()
			Settings.Reset()

			If grid IsNot Nothing Then
				If grid.Columns IsNot Nothing Then
					Settings.ColumnSettings.Clear()

					For Each column As GridViewColumn In grid.Columns
						If TypeOf column Is GridViewDataColumn Then
							Dim dataColumn As GridViewDataColumn = DirectCast(column, GridViewDataColumn)

							Dim setting As New ColumnSetting()
							setting.PropertyName = dataColumn.DataMemberBinding.Path.Path
							setting.UniqueName = dataColumn.UniqueName
							setting.Header = dataColumn.Header
							setting.Width = dataColumn.ActualWidth
							setting.DisplayIndex = dataColumn.DisplayIndex

							Settings.ColumnSettings.Add(setting)
						End If
					Next
				End If

				If grid.FilterDescriptors IsNot Nothing Then
					Settings.FilterSettings.Clear()

					For Each cfd As IColumnFilterDescriptor In grid.FilterDescriptors.OfType(Of IColumnFilterDescriptor)()
						Dim setting As New FilterSetting()

						setting.Filter1 = New Telerik.Windows.Data.FilterDescriptor()
						setting.Filter1.Member = cfd.FieldFilter.Filter1.Member
						setting.Filter1.[Operator] = cfd.FieldFilter.Filter1.[Operator]
						setting.Filter1.Value = cfd.FieldFilter.Filter1.Value
						setting.Filter1.MemberType = Nothing

						setting.Filter2 = New Telerik.Windows.Data.FilterDescriptor()
						setting.Filter2.Member = cfd.FieldFilter.Filter2.Member
						setting.Filter2.[Operator] = cfd.FieldFilter.Filter2.[Operator]
						setting.Filter2.Value = cfd.FieldFilter.Filter2.Value
						setting.Filter2.MemberType = Nothing

						For Each fd As Telerik.Windows.Data.FilterDescriptor In cfd.DistinctFilter.FilterDescriptors.OfType(Of Telerik.Windows.Data.FilterDescriptor)()
							Dim newFd As New Telerik.Windows.Data.FilterDescriptor()
							newFd.Member = fd.Member
							newFd.[Operator] = fd.[Operator]
							newFd.Value = fd.Value
							newFd.MemberType = Nothing
							setting.SelectedDistinctValues.Add(newFd)
						Next

						setting.PropertyName = cfd.Column.DataMemberBinding.Path.Path

						Settings.FilterSettings.Add(setting)
					Next
				End If

				If grid.SortDescriptors IsNot Nothing Then
					Settings.SortSettings.Clear()

					For Each d As Telerik.Windows.Data.SortDescriptor In grid.SortDescriptors
						Dim setting As New SortSetting()

						setting.PropertyName = d.Member
						setting.SortDirection = d.SortDirection

						Settings.SortSettings.Add(setting)
					Next
				End If

				If grid.GroupDescriptors IsNot Nothing Then
					Settings.GroupSettings.Clear()

					For Each d As Telerik.Windows.Data.GroupDescriptor In grid.GroupDescriptors
						Dim setting As New GroupSetting()

						setting.PropertyName = d.Member
						setting.SortDirection = d.SortDirection
						setting.DisplayContent = d.DisplayContent

						Settings.GroupSettings.Add(setting)
					Next
				End If

				Settings.FrozenColumnCount = grid.FrozenColumnCount
			End If

			Settings.Save()
		End Sub

		Private Sub Attach()
			If Me.grid IsNot Nothing Then
				AddHandler Me.grid.LayoutUpdated, New EventHandler(AddressOf LayoutUpdated)
				AddHandler Me.grid.Loaded, AddressOf Loaded
				AddHandler Application.Current.[Exit], AddressOf Current_Exit
			End If
		End Sub

		Private Sub Current_Exit(sender As Object, e As EventArgs)
			SaveState()
		End Sub

		Private Sub Loaded(sender As Object, e As EventArgs)
			LoadState()
		End Sub

		Private Sub LayoutUpdated(sender As Object, e As EventArgs)
			If grid.Parent Is Nothing Then
				SaveState()
			End If
		End Sub

		Private gridViewApplicationSettings As RadGridViewApplicationSettings = Nothing

		Protected Overridable Function CreateRadGridViewApplicationSettingsInstance() As RadGridViewApplicationSettings
			Return New RadGridViewApplicationSettings(Me)
		End Function

		Protected ReadOnly Property Settings() As RadGridViewApplicationSettings
			Get
				If gridViewApplicationSettings Is Nothing Then
					gridViewApplicationSettings = CreateRadGridViewApplicationSettingsInstance()
				End If
				Return gridViewApplicationSettings
			End Get
		End Property
	End Class

	Public Class PropertySetting
		Private _PropertyName As String
		Public Property PropertyName() As String
			Get
				Return _PropertyName
			End Get
			Set
				_PropertyName = value
			End Set
		End Property
	End Class

	Public Class SortSetting
		Inherits PropertySetting
		Private _SortDirection As ListSortDirection
		Public Property SortDirection() As ListSortDirection
			Get
				Return _SortDirection
			End Get
			Set
				_SortDirection = value
			End Set
		End Property
	End Class

	Public Class GroupSetting
		Inherits PropertySetting
		Private _DisplayContent As Object
		Public Property DisplayContent() As Object
			Get
				Return _DisplayContent
			End Get
			Set
				_DisplayContent = value
			End Set
		End Property

		Private _SortDirection As System.Nullable(Of ListSortDirection)
		Public Property SortDirection() As System.Nullable(Of ListSortDirection)
			Get
				Return _SortDirection
			End Get
			Set
				_SortDirection = value
			End Set
		End Property
	End Class

	Public Class FilterSetting
		Inherits PropertySetting
		Private _SelectedDistinctValues As List(Of Telerik.Windows.Data.FilterDescriptor)
		Public ReadOnly Property SelectedDistinctValues() As List(Of Telerik.Windows.Data.FilterDescriptor)
			Get
				If _SelectedDistinctValues Is Nothing Then
					_SelectedDistinctValues = New List(Of Telerik.Windows.Data.FilterDescriptor)()
				End If
				Return _SelectedDistinctValues
			End Get
		End Property

		Private _Filter1 As Telerik.Windows.Data.FilterDescriptor
		Public Property Filter1() As Telerik.Windows.Data.FilterDescriptor
			Get
				Return _Filter1
			End Get
			Set
				_Filter1 = value
			End Set
		End Property

		Private _Filter2 As Telerik.Windows.Data.FilterDescriptor
		Public Property Filter2() As Telerik.Windows.Data.FilterDescriptor
			Get
				Return _Filter2
			End Get
			Set
				_Filter2 = value
			End Set
		End Property
	End Class

	Public Class ColumnSetting
		Inherits PropertySetting
		Private _UniqueName As String
		Public Property UniqueName() As String
			Get
				Return _UniqueName
			End Get
			Set
				_UniqueName = value
			End Set
		End Property

		Private _Header As Object
		Public Property Header() As Object
			Get
				Return _Header
			End Get
			Set
				_Header = value
			End Set
		End Property

		Private _Width As System.Nullable(Of Double)
		Public Property Width() As System.Nullable(Of Double)
			Get
				Return _Width
			End Get
			Set
				_Width = value
			End Set
		End Property

		Private _DisplayIndex As System.Nullable(Of Integer)
		Public Property DisplayIndex() As System.Nullable(Of Integer)
			Get
				Return _DisplayIndex
			End Get
			Set
				_DisplayIndex = value
			End Set
		End Property
	End Class
End Namespace
